src/openforms/submissions/models/submission_step.py

```diff
@@ -1,8 +1,83 @@
 import uuid
+from copy import deepcopy
 
+from django.core import serializers
+from django.core.serializers.json import DjangoJSONEncoder
 from django.db import models
 from django.utils.translation import gettext_lazy as _
 
+from frozendict import frozendict
+
+from openforms.forms.models import FormDefinition, FormStep
+
+
+def _make_frozen(obj):
+    if not isinstance(obj, (dict, list)):
+        return obj
+
+    elif isinstance(obj, list):
+        return tuple([_make_frozen(item) for item in obj])
+
+    retval = frozendict()
+    for key, value in obj.items():
+        if isinstance(value, dict):
+            frozen_value = _make_frozen(value)
+        elif isinstance(value, list):
+            frozen_value = tuple([_make_frozen(item) for item in value])
+        else:
+            frozen_value = value
+        retval = retval.set(key, frozen_value)
+    return retval
+
+
+def _make_unfrozen(obj):
+    if isinstance(obj, frozendict):
+        return {key: _make_unfrozen(value) for key, value in obj.items()}
+
+    elif isinstance(obj, tuple):
+        return [_make_unfrozen(item) for item in obj]
+
+    return obj
+
+
+# Note that this exists because of data loss bug #2135 - when a FormStep is being
+# deleted in the form designer (or during import/export), we do not want to destroy
```

Comment:
> Does this also happen in import export (I thought it was only deleting the form step or restoring a form from an older version)? During the export all the form resources are turned to JSON (i don't think anything is being deleted?) and during the import, the form should be new, so there is no submission step data that could be associated with it?

I'll have to double check that for exactly which mechanism is in play, but it might be that you're importing over an existing form and updating it that way? The original issue reported it being a problem with imports, though I was able to reproduce with a simplified case of deleting the form step.

Reply:
so, when restoring old versions of a form, this flow is hit - there is an existing form instance, which makes you hit this codepath (via https://github.com/open-formulieren/open-forms/blob/master/src/openforms/forms/models/form.py#L434):

https://github.com/open-formulieren/open-forms/blob/stable/1.1.x/src/openforms/forms/utils.py#L126

I'll add a testcase that essentially operates on `Form.restore_old_version` to ensure this is also covered by the patch.